### PR TITLE
[KYUUBI #4994][REST] Support listing all engines

### LINKED ISSUE #4994: [FEATURE] Support listing all engines
### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

### Search before asking

- [X] I have searched in the [issues](https://github.com/apache/kyuubi/issues?q=is%3Aissue) and found no similar issues.

### Describe the feature

The `ListEngines` function doesn't support listing all active engines. We should support it so that an administrator can manage the cluster resources.

This could be a new RESTful admin GET API. 

### Motivation

_No response_

### Describe the solution

_No response_

### Additional context

_No response_

### Are you willing to submit PR?

- [ ] Yes. I would be willing to submit a PR with guidance from the Kyuubi community to improve.
- [X] No. I cannot submit a PR at this time.

### PR DESCRIPTION
**Pull request by lsm1**
### _Why are the changes needed?_

close #4994

### _How was this patch tested?_
- [x] Add some test cases that check the changes thoroughly including negative and positive cases if possible

- [ ] Add screenshots for manual tests if appropriate

- [ ] [Run test](https://kyuubi.readthedocs.io/en/master/contributing/code/testing.html#running-tests) locally before make a pull request

### CODE REVIEW

**Comment by pan3793** [comment]
Seems it's not required to introduce a new API, just making the query conditions `type`, `sharelevel`, `subdomain` optional is enough?

File: kyuubi-ctl/src/main/scala/org/apache/kyuubi/ctl/cli/AdminControlCliArguments.scala [hunk 1]
```diff
@@ -37,6 +37,7 @@ class AdminControlCliArguments(args: Seq[String], env: Map[String, String] = sys
     cliConfig.action match {
       case ControlAction.LIST => cliConfig.resource match {
           case ControlObject.ENGINE => new AdminListEngineCommand(cliConfig)
```
Comment (cxzl25) [inline kyuubi-ctl/src/main/scala/org/apache/kyuubi/ctl/cli/AdminControlCliArguments.scala:39]: How about adding an optional parameter to `ListEngineCommand`?
```
bin/kyuubi-admin list engine [options]
```

File: kyuubi-server/src/main/scala/org/apache/kyuubi/server/api/v1/AdminResource.scala [hunk 2]
```diff
@@ -267,7 +267,44 @@ private[v1] class AdminResource extends ApiRequestContext with Logging {
       @QueryParam("type") engineType: String,
       @QueryParam("sharelevel") shareLevel: String,
       @QueryParam("subdomain") subdomain: String,
-      @QueryParam("hive.server2.proxy.user") hs2ProxyUser: String): Seq[Engine] = {
+      @QueryParam("hive.server2.proxy.user") hs2ProxyUser: String,
+      @QueryParam("allengine") allengine: Boolean): Seq[Engine] = {
```
Comment (cxzl25) [inline kyuubi-server/src/main/scala/org/apache/kyuubi/server/api/v1/AdminResource.scala:271]:
```suggestion
      @QueryParam("allengine") allEngine: Boolean): Seq[Engine] = {
```

File: kyuubi-server/src/main/scala/org/apache/kyuubi/server/api/v1/AdminResource.scala [hunk 3]
```diff
@@ -267,7 +267,44 @@ private[v1] class AdminResource extends ApiRequestContext with Logging {
...
+    if (allengine) {
+      val engines = ListBuffer[Engine]()
```
Comment (cxzl25) [inline kyuubi-server/src/main/scala/org/apache/kyuubi/server/api/v1/AdminResource.scala:273]: When listing all engines, we need to check whether the user is an administrator.

https://github.com/apache/kyuubi/blob/beea6d5fce94043c29ed2681fc65e704f523a916/kyuubi-server/src/main/scala/org/apache/kyuubi/server/api/v1/AdminResource.scala#L331-L334

**Review by cxzl25 — APPROVED** [review]

File: kyuubi-server/src/main/scala/org/apache/kyuubi/server/api/v1/AdminResource.scala [hunk 4]
```diff
@@ -267,7 +267,51 @@ private[v1] class AdminResource extends ApiRequestContext with Logging {
       @QueryParam("type") engineType: String,
       @QueryParam("sharelevel") shareLevel: String,
       @QueryParam("subdomain") subdomain: String,
-      @QueryParam("hive.server2.proxy.user") hs2ProxyUser: String): Seq[Engine] = {
+      @QueryParam("hive.server2.proxy.user") hs2ProxyUser: String,
+      @QueryParam("allengine") allEngine: Boolean): Seq[Engine] = {
```
Comment (pan3793) [inline kyuubi-server/src/main/scala/org/apache/kyuubi/server/api/v1/AdminResource.scala:271]:
```suggestion
      @QueryParam("all") all: Boolean): Seq[Engine] = {
```

File: docs/tools/kyuubi-admin.rst [hunk 5]
```diff
@@ -97,6 +97,8 @@ Usage: ``bin/kyuubi-admin list engine [options]``
      - The subdomain for the share level of an engine. If not specified, it will read the configuration item kyuubi.engine.share.level.subdomain from kyuubi-defaults.conf.
    * - --hs2ProxyUser
      - The proxy user to impersonate. When specified, it will list engines for the hs2ProxyUser.
+   * - -ae --all-engine
```
Comment (pan3793) [inline docs/tools/kyuubi-admin.rst:100]:
```suggestion
   * - -ae --all
```

File: docs/tools/kyuubi-admin.rst [hunk 6]
```diff
@@ -97,7 +97,7 @@ Usage: ``bin/kyuubi-admin list engine [options]``
      - The subdomain for the share level of an engine. If not specified, it will read the configuration item kyuubi.engine.share.level.subdomain from kyuubi-defaults.conf.
    * - --hs2ProxyUser
      - The proxy user to impersonate. When specified, it will list engines for the hs2ProxyUser.
-   * - -ae --all-engine
+   * - -ae --all
```
Comment (pan3793) [inline docs/tools/kyuubi-admin.rst:100]: `-ae` could be `-a` then?

**Review by pan3793 — APPROVED** [review]
LGTM, rebase master should fix the CI

**Comment by cxzl25** [comment]
Thanks, merged to master/1.8.

**Comment by pan3793** [comment]
The previous thought is not mature, this change makes things messer, I propose to revert this change and re-design how can we improve the engine listing API.

**Comment by cxzl25** [comment]
> I propose to revert this change

OK

**Comment by pan3793** [comment]
reverted via https://github.com/apache/kyuubi/commit/8980e0752d86d33eee3e915c37902e88886c8f2b (master) and https://github.com/apache/kyuubi/commit/86b9f1c2ffcdbac40f5c376149f0dd42df8806c6 (branch-1.8)

**Pull request closed without merging**